# DX 704 Week 1 Project

This week's project will build a portfolio risk and return model, and make investing recommendations for hypothetical clients.
You will collect historical data, estimate returns and risks, construct efficient frontier portfolios, and sanity check the certainty of the maximum return portfolio.

The full project description and a template notebook are available on GitHub at the following link.

https://github.com/bu-cds-dx704/dx704-project-01


Feel free to use optimization tools or libraries (such as CVXOPT or scipy.optimize) to perform any calculations required for this mini project.

### Example Code

You may find it helpful to refer to these GitHub repositories of Jupyter notebooks for example code.

* https://github.com/bu-cds-omds/dx601-examples
* https://github.com/bu-cds-omds/dx602-examples
* https://github.com/bu-cds-omds/dx603-examples
* https://github.com/bu-cds-omds/dx704-examples

Any calculations demonstrated in code examples or videos may be found in these notebooks, and you are allowed to copy this example code in your homework answers.

## Part 1: Collect Data

Collect historical monthly price data for the last 24 months covering 6 different stocks.
The data should cover 24 consecutive months including the last month that ended before this week's material was released on Blackboard.
To be clear, if a month ends between the Blackboard release and submitting your project, you do not need to add that month.

The six different stocks must include AAPL, SPY and TSLA.
At least one of the remaining 3 tickers must start with the same letter as your last name (e.g. professor Considine could use COIN).
This is to encourage diversity in what stocks you analyze; if you discuss this project with classmates, please make sure that you pick different tickers to differentiate your work.
Do not pick stocks with fewer than 24 consecutive months of price data.

In [1]:
# YOUR CHANGES HERE

import yfinance as yf
import pandas as pd

Save the data as a TSV file named "historical_prices.tsv" and include a header row with the column names "date" and the 6 stock ticker symbols.
The date should be the last trading day of the month, so it may not be the last day of the month.
For example, the last trading day of November 2024 was 2024-11-29.
The remaining columns should contain the adjusted closing prices of the corresponding stock tickers on that day.


In [2]:
# YOUR CHANGES HERE

tickers = ["AAPL", "SPY", "TSLA", "LMT", "JPM", "XOM"]

data = yf.download(
    tickers,
    start="2024-09-01",
    end="2026-09-01",
    auto_adjust=False
)

data.head()

[*********************100%***********************]  6 of 6 completed


Price        Adj Close                                                  \
Ticker            AAPL         JPM         LMT         SPY        TSLA   
Date                                                                     
2024-09-03  220.921921  211.417068  537.804565  539.312683  210.600006   
2024-09-04  219.017868  210.486176  544.972534  538.208740  219.410004   
2024-09-05  220.535172  208.854752  539.103516  536.899841  230.169998   
2024-09-06  218.988113  203.893204  537.245239  527.863647  210.729996   
2024-09-09  219.077377  208.067780  546.669617  533.773682  216.270004   

Price                        Close                                      ...  \
Ticker             XOM        AAPL         JPM         LMT         SPY  ...   
Date                                                                    ...   
2024-09-03  108.301743  222.770004  220.300003  567.219971  552.080017  ...   
2024-09-04  106.979271  220.850006  219.330002  574.780029  550.950012  ...   
2024-09-05  106.144516  222.380005  217.630005  568.590027  549.609985  ...   
2024-09-06  105.647423  220.820007  212.460007  566.630005  540.359985  ...   
2024-09-09  107.870300  220.910004  216.809998  576.570007  546.409973  ...   

Price             Open                                        Volume           \
Ticker             LMT         SPY        TSLA         XOM      AAPL      JPM   
Date                                                                            
2024-09-03  567.330017  560.469971  215.259995  115.839996  50190600  8956100   
2024-09-04  569.340027  550.200012  210.589996  115.290001  43840200  7389500   
2024-09-05  575.200012  550.890015  223.490005  115.070000  36615400  8067900   
2024-09-06  569.049988  549.940002  232.600006  113.589996  48423000  7777000   
2024-09-09  567.510010  544.650024  216.199997  112.870003  67180000  8935100   

Price                                               
Ticker          LMT       SPY       TSLA       XOM  
Date                                                
2024-09-03  1266900  60600100   76714200  15725100  
2024-09-04   941700  47224900   80651800  12620600  
2024-09-05   841000  44264300  119355000  17200800  
2024-09-06   976700  68493800  112177000  13467700  
2024-09-09   810200  40445800   67443500  21645300  

[5 rows x 36 columns]

Submit "historical_prices.tsv" in Gradescope.

In [3]:
prices = data["Adj Close"]

prices.head()

Ticker,AAPL,JPM,LMT,SPY,TSLA,XOM
Date,,,,,,
2024-09-03,220.921921,211.417068,537.804565,539.312683,210.600006,108.301743
2024-09-04,219.017868,210.486176,544.972534,538.208740,219.410004,106.979271
2024-09-05,220.535172,208.854752,539.103516,536.899841,230.169998,106.144516
2024-09-06,218.988113,203.893204,537.245239,527.863647,210.729996,105.647423
2024-09-09,219.077377,208.067780,546.669617,533.773682,216.270004,107.870300


In [4]:
monthly_prices = prices.resample("ME").last()

monthly_prices

Ticker,AAPL,JPM,LMT,SPY,TSLA,XOM
Date,,,,,,
2024-09-30,231.067062,202.357712,554.245300,562.210449,261.630005,109.943123
2024-10-31,224.035873,214.276825,517.732422,557.193542,249.850006,109.530411
2024-11-30,235.620102,241.119400,501.955353,590.420837,345.160004,111.546280
2024-12-31,248.615814,231.454163,463.629639,576.215332,403.839996,101.721199
2025-01-31,234.299698,259.426880,441.695129,591.690308,404.600006,101.021454
2025-02-28,240.361572,256.854919,429.692657,584.178894,292.980011,106.218452
2025-03-31,220.772079,238.074875,429.346680,551.628906,259.160004,113.469498
2025-04-30,211.200958,238.877243,459.180176,546.846191,282.160004,100.780144
2025-05-31,199.883957,257.802277,463.630219,581.212708,346.459991,98.501991


In [5]:
print(monthly_prices.shape)
print(monthly_prices.index.min())
print(monthly_prices.index.max())

(24, 6)
2024-09-30 00:00:00
2026-08-31 00:00:00


In [6]:
monthly_prices.index = monthly_prices.index.strftime("%Y-%m")

monthly_prices = monthly_prices.round(2)

monthly_prices

Ticker,AAPL,JPM,LMT,SPY,TSLA,XOM
Date,,,,,,
2024-09,231.07,202.36,554.25,562.21,261.63,109.94
2024-10,224.04,214.28,517.73,557.19,249.85,109.53
2024-11,235.62,241.12,501.96,590.42,345.16,111.55
2024-12,248.62,231.45,463.63,576.22,403.84,101.72
2025-01,234.30,259.43,441.70,591.69,404.60,101.02
2025-02,240.36,256.85,429.69,584.18,292.98,106.22
2025-03,220.77,238.07,429.35,551.63,259.16,113.47
2025-04,211.20,238.88,459.18,546.85,282.16,100.78
2025-05,199.88,257.80,463.63,581.21,346.46,98.50


In [7]:
check = pd.read_csv("historical_prices.tsv", sep="\t")

print(check)
print("\nShape:", check.shape)
print("\nColumns:", check.columns.tolist())

          date        AAPL         SPY        TSLA         LMT         JPM  \
0   2024-09-30  231.067078  562.210449  261.630005  554.245178  202.357727   
1   2024-10-31  224.035889  557.193481  249.850006  517.732361  214.276840   
2   2024-11-29  235.620117  590.420898  345.160004  501.955231  241.119385   
3   2024-12-31  248.615784  576.215271  403.839996  463.629608  231.454147   
4   2025-01-31  234.299698  591.690369  404.600006  441.695160  259.426910   
5   2025-02-28  240.361588  584.178955  292.980011  429.692657  256.854919   
6   2025-03-31  220.772095  551.628967  259.160004  429.346649  238.074875   
7   2025-04-30  211.200943  546.846252  282.160004  459.180206  238.877258   
8   2025-05-30  199.883942  581.212830  346.459991  463.630219  257.802307   
9   2025-06-30  204.183151  611.079041  317.660004  448.204285  283.104004   
10  2025-07-31  206.571625  625.153198  308.269989  407.403870  290.679108   
11  2025-08-29  231.285629  637.981018  333.869995  440.936462  

In [8]:
monthly_prices.to_csv(
    "historical_prices.tsv",
    sep="\t"
)

## Part 2: Calculate Historical Asset Returns

Calculate the historical asset returns based on the price data that you previously collected.

In [9]:
# Get adjusted closing prices
prices = data["Adj Close"].copy()

# Put columns in required order
tickers = ["AAPL", "SPY", "TSLA", "LMT", "JPM", "XOM"]
prices = prices[tickers]

# Select the actual last trading observation in each month
monthly_prices = prices.groupby(prices.index.to_period("M")).tail(1)

# Make sure the index is named exactly "date"
monthly_prices.index.name = "date"

# Format as YYYY-MM-DD while preserving actual trading dates
monthly_prices.index = monthly_prices.index.strftime("%Y-%m-%d")

# Save as TSV
monthly_prices.to_csv(
    "historical_prices.tsv",
    sep="\t"
)

monthly_prices

Ticker,AAPL,SPY,TSLA,LMT,JPM,XOM
date,,,,,,
2024-09-30,231.067062,562.210449,261.630005,554.245300,202.357712,109.943123
2024-10-31,224.035873,557.193542,249.850006,517.732422,214.276825,109.530411
2024-11-29,235.620102,590.420837,345.160004,501.955353,241.119400,111.546280
2024-12-31,248.615814,576.215332,403.839996,463.629639,231.454163,101.721199
2025-01-31,234.299698,591.690308,404.600006,441.695129,259.426880,101.021454
2025-02-28,240.361572,584.178894,292.980011,429.692657,256.854919,106.218452
2025-03-31,220.772079,551.628906,259.160004,429.346680,238.074875,113.469498
2025-04-30,211.200958,546.846191,282.160004,459.180176,238.877243,100.780144
2025-05-30,199.883957,581.212708,346.459991,463.630219,257.802277,98.501991


In [10]:
check = pd.read_csv(
    "historical_prices.tsv",
    sep="\t"
)

print(check.head())
print()
print(check.columns.tolist())
print("Shape:", check.shape)

         date        AAPL         SPY        TSLA         LMT         JPM  \
0  2024-09-30  231.067062  562.210449  261.630005  554.245300  202.357712   
1  2024-10-31  224.035873  557.193542  249.850006  517.732422  214.276825   
2  2024-11-29  235.620102  590.420837  345.160004  501.955353  241.119400   
3  2024-12-31  248.615814  576.215332  403.839996  463.629639  231.454163   
4  2025-01-31  234.299698  591.690308  404.600006  441.695129  259.426880   

          XOM  
0  109.943123  
1  109.530411  
2  111.546280  
3  101.721199  
4  101.021454  

['date', 'AAPL', 'SPY', 'TSLA', 'LMT', 'JPM', 'XOM']
Shape: (24, 7)


In [11]:
# YOUR CHANGES HERE

import pandas as pd

prices = pd.read_csv(
    "historical_prices.tsv",
    sep="\t",
    parse_dates=["date"]
)

prices.head()

,date,AAPL,SPY,TSLA,LMT,JPM,XOM
0,2024-09-30,231.067062,562.210449,261.630005,554.245300,202.357712,109.943123
1,2024-10-31,224.035873,557.193542,249.850006,517.732422,214.276825,109.530411
2,2024-11-29,235.620102,590.420837,345.160004,501.955353,241.119400,111.546280
3,2024-12-31,248.615814,576.215332,403.839996,463.629639,231.454163,101.721199
4,2025-01-31,234.299698,591.690308,404.600006,441.695129,259.426880,101.021454


In [12]:
tickers = ["AAPL", "SPY", "TSLA", "LMT", "JPM", "XOM"]

returns = prices.copy()

returns[tickers] = prices[tickers].pct_change()

returns.head()

,date,AAPL,SPY,TSLA,LMT,JPM,XOM
0,2024-09-30,NaN,NaN,NaN,NaN,NaN,NaN
1,2024-10-31,-0.030429,-0.008924,-0.045025,-0.065879,0.058901,-0.003754
2,2024-11-29,0.051707,0.059633,0.381469,-0.030473,0.125271,0.018405
3,2024-12-31,0.055155,-0.024060,0.170008,-0.076353,-0.040085,-0.088081
4,2025-01-31,-0.057583,0.026856,0.001882,-0.047310,0.120856,-0.006879


In [13]:
returns = returns.dropna().reset_index(drop=True)

returns.head()

,date,AAPL,SPY,TSLA,LMT,JPM,XOM
0,2024-10-31,-0.030429,-0.008924,-0.045025,-0.065879,0.058901,-0.003754
1,2024-11-29,0.051707,0.059633,0.381469,-0.030473,0.125271,0.018405
2,2024-12-31,0.055155,-0.024060,0.170008,-0.076353,-0.040085,-0.088081
3,2025-01-31,-0.057583,0.026856,0.001882,-0.047310,0.120856,-0.006879
4,2025-02-28,0.025872,-0.012695,-0.275877,-0.027174,-0.009914,0.051445


In [14]:
print("Number of return periods:", len(returns))
print("Shape:", returns.shape)
print()
print(returns.head())

Number of return periods: 23
Shape: (23, 7)

        date      AAPL       SPY      TSLA       LMT       JPM       XOM
0 2024-10-31 -0.030429 -0.008924 -0.045025 -0.065879  0.058901 -0.003754
1 2024-11-29  0.051707  0.059633  0.381469 -0.030473  0.125271  0.018405
2 2024-12-31  0.055155 -0.024060  0.170008 -0.076353 -0.040085 -0.088081
3 2025-01-31 -0.057583  0.026856  0.001882 -0.047310  0.120856 -0.006879
4 2025-02-28  0.025872 -0.012695 -0.275877 -0.027174 -0.009914  0.051445


In [15]:
print("AAPL October 2024 return:", returns.loc[0, "AAPL"])

AAPL October 2024 return: -0.030429213460738125


Save the data as a TSV file named "historical_returns.tsv" and include a header row with the column names "date" and the 6 stock ticker symbols.
Each row should have the date at the end of the month and the corresponding *relative* price changes.
For example, if the previous price was \$100 and the new price is \$110, the return value should be 0.10.
There should only be 23 rows of data in this file, since they are computed as the differences of 24 prices.

In [16]:
# YOUR CHANGES HERE

returns.to_csv(
    "historical_returns.tsv",
    sep="\t",
    index=False
)

In [17]:
check_returns = pd.read_csv(
    "historical_returns.tsv",
    sep="\t"
)

print(check_returns.head())
print()
print("Shape:", check_returns.shape)
print("Columns:", check_returns.columns.tolist())
print()
print("Missing values:")
print(check_returns.isna().sum())

         date      AAPL       SPY      TSLA       LMT       JPM       XOM
0  2024-10-31 -0.030429 -0.008924 -0.045025 -0.065879  0.058901 -0.003754
1  2024-11-29  0.051707  0.059633  0.381469 -0.030473  0.125271  0.018405
2  2024-12-31  0.055155 -0.024060  0.170008 -0.076353 -0.040085 -0.088081
3  2025-01-31 -0.057583  0.026856  0.001882 -0.047310  0.120856 -0.006879
4  2025-02-28  0.025872 -0.012695 -0.275877 -0.027174 -0.009914  0.051445

Shape: (23, 7)
Columns: ['date', 'AAPL', 'SPY', 'TSLA', 'LMT', 'JPM', 'XOM']

Missing values:
date    0
AAPL    0
SPY     0
TSLA    0
LMT     0
JPM     0
XOM     0
dtype: int64


Submit "historical_returns.tsv" in Gradescope.

## Part 3: Estimate Returns

Estimate the expected returns for each asset using the previously calculated return data.
Just compute the average (mean) return for each asset over your data set; do not use other estimators that have been mentioned.
This will serve as your estimate of expected return for each asset.

In [18]:
# YOUR CHANGES HERE

import pandas as pd

returns = pd.read_csv(
    "historical_returns.tsv",
    sep="\t"
)

returns.head()

,date,AAPL,SPY,TSLA,LMT,JPM,XOM
0,2024-10-31,-0.030429,-0.008924,-0.045025,-0.065879,0.058901,-0.003754
1,2024-11-29,0.051707,0.059633,0.381469,-0.030473,0.125271,0.018405
2,2024-12-31,0.055155,-0.024060,0.170008,-0.076353,-0.040085,-0.088081
3,2025-01-31,-0.057583,0.026856,0.001882,-0.047310,0.120856,-0.006879
4,2025-02-28,0.025872,-0.012695,-0.275877,-0.027174,-0.009914,0.051445


In [19]:
tickers = ["AAPL", "SPY", "TSLA", "LMT", "JPM", "XOM"]

expected_returns = returns[tickers].mean()

print(expected_returns)

AAPL    0.015696
SPY     0.014246
TSLA    0.027043
LMT     0.004359
JPM     0.026345
XOM     0.019123
dtype: float64


Save the estimated returns in a TSV file named "estimated_returns.tsv" and include a header row with the column names "asset" and "estimated_return".

In [20]:
# YOUR CHANGES HERE

estimated_returns_df = expected_returns.reset_index()

estimated_returns_df.columns = ["asset", "estimated_return"]

estimated_returns_df.to_csv(
    "estimated_returns.tsv",
    sep="\t",
    index=False
)

estimated_returns_df

,asset,estimated_return
0,AAPL,0.015696
1,SPY,0.014246
2,TSLA,0.027043
3,LMT,0.004359
4,JPM,0.026345
5,XOM,0.019123


In [21]:
check_estimates = pd.read_csv(
    "estimated_returns.tsv",
    sep="\t"
)

print(check_estimates)
print()
print("Shape:", check_estimates.shape)
print("Columns:", check_estimates.columns.tolist())

  asset  estimated_return
0  AAPL          0.015696
1   SPY          0.014246
2  TSLA          0.027043
3   LMT          0.004359
4   JPM          0.026345
5   XOM          0.019123

Shape: (6, 2)
Columns: ['asset', 'estimated_return']


Submit "estimated_returns.tsv" in Gradescope.

## Part 4: Estimate Risk

Estimate the covariance matrix for the asset returns to understand how the assets move together.

In [22]:
# YOUR CHANGES HERE
returns = pd.read_csv(
    "historical_returns.tsv",
    sep="\t"
)

tickers = ["AAPL", "SPY", "TSLA", "LMT", "JPM", "XOM"]


In [23]:
covariance_matrix = returns[tickers].cov()

covariance_matrix

,AAPL,SPY,TSLA,LMT,JPM,XOM
AAPL,0.004014,0.001053,0.002829,0.000046,-0.000234,-0.000630
SPY,0.001053,0.001380,0.002818,-0.000311,0.001095,-0.000822
TSLA,0.002829,0.002818,0.026124,-0.000388,0.001768,-0.005484
LMT,0.000046,-0.000311,-0.000388,0.009227,-0.001197,0.003327
JPM,-0.000234,0.001095,0.001768,-0.001197,0.003181,-0.000785
XOM,-0.000630,-0.000822,-0.005484,0.003327,-0.000785,0.005146


Save the estimated covariances to a TSV file named "estimated_covariance.tsv".
The header row should have a blank column name followed by the names of the assets.
Each data row should start with the name of an asset for that row, and be followed by the individual covariances corresponding to that row and column's assets.
(This is the format of pandas's `to_csv` method with `sep="\t"` when used on a covariance matrix as computed in the examples.)

In [24]:
# YOUR CHANGES HERE
covariance_matrix.to_csv(
    "estimated_covariance.tsv",
    sep="\t"
)

Submit "estimated_covariance.tsv" in Gradescope.

## Part 5: Construct the Maximum Return Portfolio

Compute the maximum return portfolio based on your previously estimated risks and returns.

In [25]:
# YOUR CHANGES HERE

estimated_returns = pd.read_csv(
    "estimated_returns.tsv",
    sep="\t"
)

estimated_returns

,asset,estimated_return
0,AAPL,0.015696
1,SPY,0.014246
2,TSLA,0.027043
3,LMT,0.004359
4,JPM,0.026345
5,XOM,0.019123


In [26]:
max_return_asset = estimated_returns.loc[
    estimated_returns["estimated_return"].idxmax()
]

print(max_return_asset)

asset                   TSLA
estimated_return    0.027043
Name: 2, dtype: object


In [27]:
tickers = ["AAPL", "SPY", "TSLA", "LMT", "JPM", "XOM"]

max_return_portfolio = pd.Series(
    0.0,
    index=tickers
)

max_return_portfolio[max_return_asset["asset"]] = 1.0

print(max_return_portfolio)

AAPL    0.0
SPY     0.0
TSLA    1.0
LMT     0.0
JPM     0.0
XOM     0.0
dtype: float64


In [28]:
portfolio_return = (
    max_return_portfolio
    * estimated_returns.set_index("asset")["estimated_return"]
).sum()

print("Portfolio expected return:", portfolio_return)

Portfolio expected return: 0.0270432456452091


In [29]:
covariance_matrix = pd.read_csv(
    "estimated_covariance.tsv",
    sep="\t",
    index_col=0
)

portfolio_variance = (
    max_return_portfolio.values
    @ covariance_matrix.loc[tickers, tickers].values
    @ max_return_portfolio.values
)

portfolio_std = portfolio_variance ** 0.5

print("Portfolio variance:", portfolio_variance)
print("Portfolio standard deviation:", portfolio_std)

Portfolio variance: 0.026124211134687
Portfolio standard deviation: 0.1616298584256232


Save the maximum return portfolio in a TSV file named "maximum_return.tsv".
The header row should have two columns, "asset" and "allocation".
The allocation values should sum up to one.


In [30]:
# YOUR CHANGES HERE

import pandas as pd

estimated_returns = pd.read_csv(
    "estimated_returns.tsv",
    sep="\t"
)

estimated_returns

,asset,estimated_return
0,AAPL,0.015696
1,SPY,0.014246
2,TSLA,0.027043
3,LMT,0.004359
4,JPM,0.026345
5,XOM,0.019123


In [31]:
max_asset = estimated_returns.loc[
    estimated_returns["estimated_return"].idxmax(),
    "asset"
]

print("Highest expected return asset:", max_asset)

Highest expected return asset: TSLA


In [32]:
maximum_return = pd.DataFrame({
    "asset": estimated_returns["asset"],
    "allocation": 0.0
})

maximum_return.loc[
    maximum_return["asset"] == max_asset,
    "allocation"
] = 1.0

maximum_return

,asset,allocation
0,AAPL,0.0
1,SPY,0.0
2,TSLA,1.0
3,LMT,0.0
4,JPM,0.0
5,XOM,0.0


In [33]:
print("Total allocation:", maximum_return["allocation"].sum())

Total allocation: 1.0


In [34]:
maximum_return.to_csv(
    "maximum_return.tsv",
    sep="\t",
    index=False
)

In [35]:
check = pd.read_csv(
    "maximum_return.tsv",
    sep="\t"
)

print(check)
print()
print("Columns:", check.columns.tolist())
print("Shape:", check.shape)
print("Allocation sum:", check["allocation"].sum())

  asset  allocation
0  AAPL         0.0
1   SPY         0.0
2  TSLA         1.0
3   LMT         0.0
4   JPM         0.0
5   XOM         0.0

Columns: ['asset', 'allocation']
Shape: (6, 2)
Allocation sum: 1.0


Submit "maximum_return.tsv" in Gradescope.

## Part 6: Construct the Minimum Risk Portfolio

Compute the minimum risk portfolio based on your previously estimated risks.

In [36]:
# YOUR CHANGES HERE

covariance_matrix = pd.read_csv(
    "estimated_covariance.tsv",
    sep="\t",
    index_col=0
)

tickers = ["AAPL", "SPY", "TSLA", "LMT", "JPM", "XOM"]

In [37]:
def portfolio_variance(weights, covariance_matrix):
    return weights @ covariance_matrix.values @ weights

In [38]:
import numpy as np

n_assets = len(tickers)

initial_weights = np.repeat(1 / n_assets, n_assets)

constraints = {
    "type": "eq",
    "fun": lambda weights: np.sum(weights) - 1
}

bounds = [(0, 1) for _ in range(n_assets)]

In [40]:
from scipy.optimize import minimize

result = minimize(
    portfolio_variance,
    initial_weights,
    args=(covariance_matrix.loc[tickers, tickers],),
    method="SLSQP",
    bounds=bounds,
    constraints=constraints
)

In [46]:
minimum_risk = pd.DataFrame({
    "asset": tickers,
    "allocation": result.x
})

print("Optimization successful:", result.success)
print("Allocation sum:", minimum_risk["allocation"].sum())

minimum_risk

Optimization successful: True
Allocation sum: 1.0


,asset,allocation
0,AAPL,0.198694
1,SPY,0.252294
2,TSLA,0.010909
3,LMT,0.046333
4,JPM,0.265080
5,XOM,0.226691


Save the minimum risk portfolio in a TSV file named "minimum_risk.tsv".
The header row should have two columns, "asset" and "allocation".
The allocation values should sum up to one.


In [47]:
# YOUR CHANGES HERE

minimum_risk.to_csv(
    "minimum_risk.tsv",
    sep="\t",
    index=False
)

In [48]:
check_min_risk = pd.read_csv(
    "minimum_risk.tsv",
    sep="\t"
)

print(check_min_risk)
print()
print("Columns:", check_min_risk.columns.tolist())
print("Allocation sum:", check_min_risk["allocation"].sum())

  asset  allocation
0  AAPL    0.198694
1   SPY    0.252294
2  TSLA    0.010909
3   LMT    0.046333
4   JPM    0.265080
5   XOM    0.226691

Columns: ['asset', 'allocation']
Allocation sum: 0.9999999999999996


Submit "minimum_risk.tsv" in Gradescope.

## Part 7: Build Efficient Frontier Portfolios

Compute 101 portfolios along the mean-variance efficient frontier with evenly spaced estimated returns.
The first portfolio should be the minimum risk portfolio from part 4, and the last portfolio should be the maximum return portfolio from part 3.
The estimated return of each portfolio should be higher than the previous by one percent of the difference between the first and last portfolios.
That is, the estimated return of the portfolios should be similar to `np.linspace(min_risk_return, max_return, 101)`.


In [49]:
# YOUR CHANGES HERE
import pandas as pd
import numpy as np
from scipy.optimize import minimize

estimated_returns = pd.read_csv(
    "estimated_returns.tsv",
    sep="\t"
).set_index("asset")["estimated_return"]

covariance_matrix = pd.read_csv(
    "estimated_covariance.tsv",
    sep="\t",
    index_col=0
)

minimum_risk = pd.read_csv(
    "minimum_risk.tsv",
    sep="\t"
)

maximum_return = pd.read_csv(
    "maximum_return.tsv",
    sep="\t"
)

tickers = ["AAPL", "SPY", "TSLA", "LMT", "JPM", "XOM"]


In [51]:
min_weights = minimum_risk.set_index("asset").loc[tickers, "allocation"].values
max_weights = maximum_return.set_index("asset").loc[tickers, "allocation"].values

mean_returns = estimated_returns.loc[tickers].values

min_risk_return = min_weights @ mean_returns
max_return = max_weights @ mean_returns

print("Minimum-risk portfolio return:", min_risk_return)
print("Maximum-return portfolio return:", max_return)

Minimum-risk portfolio return: 0.01852823122425687
Maximum-return portfolio return: 0.0270432456452091


In [54]:
target_returns = np.linspace(
    min_risk_return,
    max_return,
    101
)

target_returns[:5], target_returns[-5:]

(array([0.01852823, 0.01861338, 0.01869853, 0.01878368, 0.01886883]),
 array([0.02670265, 0.0267878 , 0.02687295, 0.0269581 , 0.02704325]))

In [55]:
cov_matrix = covariance_matrix.loc[tickers, tickers].values

def portfolio_variance(weights):
    return weights @ cov_matrix @ weights

In [56]:
frontier_portfolios = []

initial_weights = min_weights.copy()

bounds = [(0, 1) for _ in tickers]

for target_return in target_returns:

    constraints = [
        {
            "type": "eq",
            "fun": lambda weights: np.sum(weights) - 1
        },
        {
            "type": "eq",
            "fun": lambda weights, target=target_return:
                weights @ mean_returns - target
        }
    ]

    result = minimize(
        portfolio_variance,
        initial_weights,
        method="SLSQP",
        bounds=bounds,
        constraints=constraints
    )

    if not result.success:
        print("Optimization failed:", target_return, result.message)

    frontier_portfolios.append(result.x)

    # Use current solution as next starting point
    initial_weights = result.x

In [57]:
efficient_frontier = pd.DataFrame(
    frontier_portfolios,
    columns=tickers
)

efficient_frontier.insert(
    0,
    "estimated_return",
    target_returns
)

efficient_frontier

,estimated_return,AAPL,SPY,TSLA,LMT,JPM,XOM
0,0.018528,1.986941e-01,2.522940e-01,0.010909,4.633259e-02,2.650800e-01,2.266906e-01
1,0.018613,1.979726e-01,2.516524e-01,0.013168,4.308756e-02,2.669839e-01,2.271359e-01
2,0.018699,1.972444e-01,2.509849e-01,0.015373,3.983889e-02,2.689186e-01,2.276403e-01
3,0.018784,1.965096e-01,2.502918e-01,0.017526,3.658703e-02,2.708830e-01,2.282022e-01
4,0.018869,1.957685e-01,2.495732e-01,0.019630,3.333239e-02,2.728760e-01,2.288197e-01
...,...,...,...,...,...,...,...
96,0.026703,7.640915e-16,0.000000e+00,0.512426,0.000000e+00,4.875736e-01,0.000000e+00
97,0.026788,6.448835e-16,0.000000e+00,0.634320,3.389758e-16,3.656802e-01,0.000000e+00
98,0.026873,0.000000e+00,8.068090e-16,0.756213,0.000000e+00,2.437868e-01,2.374403e-16
99,0.026958,0.000000e+00,2.191010e-15,0.878107,0.000000e+00,1.218934e-01,3.593046e-16


In [58]:
print("Number of portfolios:", len(efficient_frontier))

print(
    "Allocation sums:",
    efficient_frontier[tickers].sum(axis=1).min(),
    "to",
    efficient_frontier[tickers].sum(axis=1).max()
)

print(
    "First return:",
    efficient_frontier["estimated_return"].iloc[0]
)

print(
    "Last return:",
    efficient_frontier["estimated_return"].iloc[-1]
)

Number of portfolios: 101
Allocation sums: 0.9999999999999996 to 1.0000000000000024
First return: 0.01852823122425687
Last return: 0.0270432456452091


In [59]:
return_steps = np.diff(
    efficient_frontier["estimated_return"]
)

print("Return step:", return_steps[0])
print("All steps equal:", np.allclose(return_steps, return_steps[0]))

Return step: 8.515014420952191e-05
All steps equal: True


Save the portfolios in a TSV file named "efficient_frontier.tsv".
The header row should have columns "index", "return", "risk", and all the asset tickers.
Each data row should have the portfolio index (0-100), the estimated return of the portfolio, the estimated standard deviation (not variance) of the portfolio, and all the asset allocations (which should sum to one).

In [60]:
# YOUR CHANGES HERE
frontier_risks = []

for weights in frontier_portfolios:
    variance = weights @ cov_matrix @ weights
    standard_deviation = np.sqrt(variance)
    frontier_risks.append(standard_deviation)


In [61]:
efficient_frontier_output = pd.DataFrame(
    frontier_portfolios,
    columns=tickers
)

efficient_frontier_output.insert(
    0,
    "risk",
    frontier_risks
)

efficient_frontier_output.insert(
    0,
    "return",
    target_returns
)

efficient_frontier_output.insert(
    0,
    "index",
    range(101)
)

efficient_frontier_output

,index,return,risk,AAPL,SPY,TSLA,LMT,JPM,XOM
0,0,0.018528,0.028000,1.986941e-01,2.522940e-01,0.010909,4.633259e-02,2.650800e-01,2.266906e-01
1,1,0.018613,0.028006,1.979726e-01,2.516524e-01,0.013168,4.308756e-02,2.669839e-01,2.271359e-01
2,2,0.018699,0.028020,1.972444e-01,2.509849e-01,0.015373,3.983889e-02,2.689186e-01,2.276403e-01
3,3,0.018784,0.028042,1.965096e-01,2.502918e-01,0.017526,3.658703e-02,2.708830e-01,2.282022e-01
4,4,0.018869,0.028071,1.957685e-01,2.495732e-01,0.019630,3.333239e-02,2.728760e-01,2.288197e-01
...,...,...,...,...,...,...,...,...,...
96,96,0.026703,0.092191,7.640915e-16,0.000000e+00,0.512426,0.000000e+00,4.875736e-01,0.000000e+00
97,97,0.026788,0.108429,6.448835e-16,0.000000e+00,0.634320,3.389758e-16,3.656802e-01,0.000000e+00
98,98,0.026873,0.125619,0.000000e+00,8.068090e-16,0.756213,0.000000e+00,2.437868e-01,2.374403e-16
99,99,0.026958,0.143420,0.000000e+00,2.191010e-15,0.878107,0.000000e+00,1.218934e-01,3.593046e-16


In [62]:
efficient_frontier_output.to_csv(
    "efficient_frontier.tsv",
    sep="\t",
    index=False
)

In [63]:
check_frontier = pd.read_csv(
    "efficient_frontier.tsv",
    sep="\t"
)

print("Shape:", check_frontier.shape)
print("Columns:", check_frontier.columns.tolist())

print(
    "Index range:",
    check_frontier["index"].min(),
    "to",
    check_frontier["index"].max()
)

allocation_sums = check_frontier[tickers].sum(axis=1)

print(
    "Allocation sums:",
    allocation_sums.min(),
    "to",
    allocation_sums.max()
)

print()
print(check_frontier.head())

Shape: (101, 9)
Columns: ['index', 'return', 'risk', 'AAPL', 'SPY', 'TSLA', 'LMT', 'JPM', 'XOM']
Index range: 0 to 100
Allocation sums: 0.9999999999999996 to 1.0000000000000024

   index    return      risk      AAPL       SPY      TSLA       LMT  \
0      0  0.018528  0.028000  0.198694  0.252294  0.010909  0.046333   
1      1  0.018613  0.028006  0.197973  0.251652  0.013168  0.043088   
2      2  0.018699  0.028020  0.197244  0.250985  0.015373  0.039839   
3      3  0.018784  0.028042  0.196510  0.250292  0.017526  0.036587   
4      4  0.018869  0.028071  0.195768  0.249573  0.019630  0.033332   

        JPM       XOM  
0  0.265080  0.226691  
1  0.266984  0.227136  
2  0.268919  0.227640  
3  0.270883  0.228202  
4  0.272876  0.228820  


In [64]:
print(check_frontier.iloc[[0, 100]])

     index    return     risk          AAPL       SPY      TSLA           LMT  \
0        0  0.018528  0.02800  1.986941e-01  0.252294  0.010909  4.633259e-02   
100    100  0.027043  0.16163  3.225501e-17  0.000000  1.000000  1.052164e-15   

              JPM       XOM  
0    2.650800e-01  0.226691  
100  2.495908e-08  0.000000  


Submit "efficient_frontier.tsv" in Gradescope.

## Part 8: Check Maximum Return Portfolio Stability

Check the stability of the maximum return portfolio by resampling the estimated risk/return model.

Repeat 1000 times -
1. Use `np.random.multivariate_normal` to generate 23 return samples using your previously estimated risks and returns.
2. Estimate the return of each asset using that resampled return history.
3. Check which asset had the highest return in those resampled estimates.

This procedure is a reduced and simplified version of the Michaud resampled efficient frontier procedure that takes uncertainty in the risk model into account.

In [66]:
# YOUR CHANGES HERE
tickers = ["AAPL", "SPY", "TSLA", "LMT", "JPM", "XOM"]

estimated_returns = pd.read_csv(
    "estimated_returns.tsv",
    sep="\t"
).set_index("asset")["estimated_return"]

covariance_matrix = pd.read_csv(
    "estimated_covariance.tsv",
    sep="\t",
    index_col=0
)

mean_vector = estimated_returns.loc[tickers].values
cov_matrix = covariance_matrix.loc[tickers, tickers].values


In [67]:
np.random.seed(42)

winner_counts = {ticker: 0 for ticker in tickers}

for _ in range(1000):

    simulated_returns = np.random.multivariate_normal(
        mean=mean_vector,
        cov=cov_matrix,
        size=23
    )

    resampled_means = simulated_returns.mean(axis=0)

    winner_index = np.argmax(resampled_means)
    winner_ticker = tickers[winner_index]

    winner_counts[winner_ticker] += 1

In [68]:
stability_results = pd.DataFrame({
    "asset": tickers,
    "times_highest_return": [winner_counts[t] for t in tickers]
})

stability_results["proportion"] = (
    stability_results["times_highest_return"] / 1000
)

stability_results

,asset,times_highest_return,proportion
0,AAPL,79,0.079
1,SPY,5,0.005
2,TSLA,412,0.412
3,LMT,48,0.048
4,JPM,249,0.249
5,XOM,207,0.207


In [69]:
print("Total simulations:", stability_results["times_highest_return"].sum())

Total simulations: 1000


Save a file "max_return_probabilities.tsv" with the distribution of highest return assets.
The header row should have columns "asset" and "probability".
There should be a data row for each asset and its sample probability of having the highest return based on those 1000 resampled estimates.


In [70]:
# YOUR CHANGES HERE

max_return_probabilities = pd.DataFrame({
    "asset": tickers,
    "probability": [winner_counts[ticker] / 1000 for ticker in tickers]
})

max_return_probabilities

,asset,probability
0,AAPL,0.079
1,SPY,0.005
2,TSLA,0.412
3,LMT,0.048
4,JPM,0.249
5,XOM,0.207


In [71]:
print(
    "Probability sum:",
    max_return_probabilities["probability"].sum()
)

Probability sum: 1.0


In [72]:
max_return_probabilities.to_csv(
    "max_return_probabilities.tsv",
    sep="\t",
    index=False
)

Submit "max_return_probabilities.tsv" in Gradescope.

In [73]:
check_probabilities = pd.read_csv(
    "max_return_probabilities.tsv",
    sep="\t"
)

print(check_probabilities)
print()
print("Shape:", check_probabilities.shape)
print("Columns:", check_probabilities.columns.tolist())
print("Probability sum:", check_probabilities["probability"].sum())

  asset  probability
0  AAPL        0.079
1   SPY        0.005
2  TSLA        0.412
3   LMT        0.048
4   JPM        0.249
5   XOM        0.207

Shape: (6, 2)
Columns: ['asset', 'probability']
Probability sum: 1.0


## Part 9: Acknowledgments

Make a file "acknowledgments.txt" documenting any outside sources or help on this project.
If you discussed this assignment with anyone, please acknowledge them here.
If you used any libraries not mentioned in this module's content, please list them with a brief explanation what you used them for.
If you used any generative AI tools, please add links to your transcripts below, and any other information that you feel is necessary to comply with the generative AI policy.
If no acknowledgments are appropriate, just write none in the file.


Submit "acknowledgments.txt" in Gradescope.

## Part 10: Code

Please submit a Jupyter notebook that can reproduce all your calculations and recreate the previously submitted files.
You do not need to provide code for data collection if you did that by manually.

Submit "project.ipynb" in Gradescope.